### PR TITLE
Refactor raw  linux clock & timex

### PR DESCRIPTION
**Pull request by folkertdev**
Preparation for extracting clock functionality into a separate crate.

### CODE REVIEW

File: statime-linux/src/clock/raw.rs [hunk 1]
```diff
@@ -71,86 +111,96 @@ impl RawLinuxClock {
...
+        let mut timex = EMPTY_TIMEX;
+        self.adjtime(&mut timex)?;
+
+        const M: f64 = 1_000_000.0;
+
+        // In struct timex, freq, ppsfreq, and stabil are ppm (parts per million) with a
+        // 16-bit fractional part, which means that a value of 1 in one of those fields
+        // actually means 2^-16 ppm, and 2^16=65536 is 1 ppm.  This is the case for both
+        // input values (in the case of freq) and output values.
+        let current_ppm = (timex.freq >> 16) as f64 + ((timex.freq & 0xffff) as f64 / 65536.0);
```
Comment (folkertdev) [inline statime-linux/src/clock/raw.rs:135]: I had to make some adjustments to the arithmetic. I have a fuzzer verifying that my new implementation matches the old one, but nonetheless this requires careful review.

File: statime-linux/src/clock/raw.rs [hunk 2]
```diff
@@ -71,86 +111,96 @@ impl RawLinuxClock {
...
-        // We do an offset with precision
-        let mut frequency_timex = Timex::new();
-
-        frequency_timex.set_status(
-            current_timex.get_status()
-                | StatusFlags::FREQHOLD // We want no automatic frequency updates
-                & !StatusFlags::PLL
-                & !StatusFlags::PPSFREQ
-                & !StatusFlags::FLL
-                & !StatusFlags::PPSTIME,
```
Comment (folkertdev) [inline statime-linux/src/clock/raw.rs:88]: we are "unsetting" bits here, but the timex is new and therefore all zeros. So all status flags except FREQHOLD never had any effect ...

File: statime-linux/src/clock/raw.rs [hunk 3]
```diff
@@ -71,86 +111,96 @@ impl RawLinuxClock {
...
+        timex.modes |= libc::ADJ_FREQUENCY | libc::ADJ_STATUS;
+
+        // NTP Kapi expects frequency adjustment in units of 2^-16 ppm
+        // but our input is in units of seconds drift per second, so convert.
+        let frequency = (ppm * 65536e6).round() as libc::c_long;
+
+        // Since Linux 2.6.26, the supplied value is clamped to the range (-32768000,
+        // +32768000). In older kernels, an EINVAL error occurs if the supplied value is
+        // out of range.
+        timex.freq = frequency.clamp(-32_768_000 + 1, 32_768_000 - 1);
```
Comment (folkertdev) [inline statime-linux/src/clock/raw.rs:178]: this clamping was performed in `Timex` before, but considered the range to be inclusive. Based on the linux documentation, the range is exclusive on both sides. (the particular value seems a bit random to me, but what do I know)

File: statime-linux/src/clock/raw.rs [hunk 4]
```diff
@@ -71,86 +111,96 @@ impl RawLinuxClock {
...
+        // This flag prevents the small frequency adjustment from being made when
+        // correcting for an ADJ_OFFSET value.
+        timex.status |= libc::STA_FREQHOLD;
+
+        // set the frequency and the status (for STA_FREQHOLD)
+        timex.modes |= libc::ADJ_FREQUENCY | libc::ADJ_STATUS;
+
+        // NTP Kapi expects frequency adjustment in units of 2^-16 ppm
+        // but our input is in units of seconds drift per second, so convert.
+        let frequency = (ppm * 65536e6).round() as libc::c_long;
```
Comment (folkertdev) [inline statime-linux/src/clock/raw.rs:173]: rounding here makes our logic equivalent to what `Fixed::to_bits` did. I don't think it matters too much in practice

File: statime-linux/src/clock/raw.rs [hunk 5]
```diff
@@ -205,37 +255,38 @@ impl RawLinuxClock {
         self.quality
     }
 
-    pub fn set_leap_seconds(&self, leap_61: bool, leap_59: bool) -> Result<(), i32> {
-        let (mut clock, _) = self.get_clock_state()?;
+    pub fn set_leap_seconds(&self, leap_61: bool, leap_59: bool) -> Result<(), Error> {
+        let leap_status = if leap_61 {
+            LeapIndicator::Leap61
+        } else if leap_59 {
```
Comment (folkertdev) [inline statime-linux/src/clock/raw.rs:261]: this assumes that we never have a forward and backward leapsecond at the same time (or close together). Seems like a fair assumption to me.

File: statime-linux/src/clock/raw.rs [hunk 6]
```diff
@@ -71,86 +107,96 @@ impl RawLinuxClock {
...
+        let mut timex = EMPTY_TIMEX;
+        self.adjtime(&mut timex)?;
+
+        const M: f64 = 1_000_000.0;
+
+        // In struct timex, freq, ppsfreq, and stabil are ppm (parts per million) with a
+        // 16-bit fractional part, which means that a value of 1 in one of those fields
+        // actually means 2^-16 ppm, and 2^16=65536 is 1 ppm.  This is the case for both
+        // input values (in the case of freq) and output values.
+        let current_ppm = (timex.freq >> 16) as f64 + ((timex.freq & 0xffff) as f64 / 65536.0);
```
Comment (davidv1992) [inline statime-linux/src/clock/raw.rs:131]: This can be simplified:
```suggestion
        let current_ppm = timex.freq as f64 / 65536.0;
```

**Review by davidv1992 — CHANGES_REQUESTED** [review]
Mostly looks sensible to me, one small but important simplification in some of the math, but other than that seems ok. Does need lab testing before merge.

**Comment by folkertdev** [comment]
before doing lab testing, i think we should look at what this function does

```rust
    fn adjust_offset(&mut self, time_offset: f64) -> Result<(), Error> {
        let mut timex = EMPTY_TIMEX;

        // set set an offset, it is in nanoseconds
        timex.modes = libc::ADJ_OFFSET | libc::ADJ_NANO;

        // Start with a seconds value of 0 and express the full time offset in nanos
        timex.time.tv_sec = time_offset as _;
        timex.time.tv_usec = (time_offset.fract() * 1_000_000_000.0) as _;

        // The nanos must not be negative. In that case the timestamp must be delivered
        // as a negative seconds with a postive nanos value
        while timex.time.tv_usec < 0 {
            timex.time.tv_sec -= 1;
            timex.time.tv_usec += 1_000_000_000;
        }

        self.adjtime(&mut timex)
    }
```

here `libc::ADJ_OFFSET` is used. However:

> ADJ_OFFSET
>       Set time offset from buf.offset.  Since Linux 2.6.26, the
>       supplied value is clamped to the range (-0.5s, +0.5s).  In
>       older kernels, an EINVAL error occurs if the supplied
>       value is out of range.

so `ADJ_OFFSET` uses the value from the `timex.offset` field. But we never set this field in our timex: we set the timex.time field. So I suspect that this function does not work and that `libc::ADJ_SETOFFSET` should be used instead

>        ADJ_SETOFFSET (since Linux 2.6.39)
>               Add buf.time to the current time.  If buf.status includes
>               the ADJ_NANO flag, then buf.time.tv_usec is interpreted as
>               a nanosecond value; otherwise it is interpreted as
>               microseconds.
> 
>               The value of buf.time is the sum of its two fields, but
>               the field buf.time.tv_usec must always be nonnegative.
>               The following example shows how to normalize a timeval
>               with nanosecond resolution.
> 
>                   while (buf.time.tv_usec < 0) {
>                       buf.time.tv_sec  -= 1;
>                       buf.time.tv_usec += 1000000000;
>                   }

**Review by davidv1992 — APPROVED** [review]

**Pull request merged**